modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java

```diff
@@ -2055,58 +2061,165 @@ public void visitRegex(ERegex userRegexNode, SemanticScope semanticScope) {
-                    + "in elasticsearch.yaml to allow them. Be careful though, regexes break out of Painless's protection against deep "
-                    + "recursion and long loops."));
+            throw userRegexNode.createError(new IllegalStateException("Regexes are disabled. Set [script.painless.regex.enabled] to "
+                    + "[limited] in elasticsearch.yaml to allow them with some protection against super long regexes. You can set it to "
+                    + "[true] to allow them without an protection against long backtracking but this usually a bad idea."));
+        }
+        switch (userRegexNode.getFlavor()) {
+            case JAVA:
+                /*
+                 * It's important for backwards compatibility that `/foo/` is
+                 * a java.util.regex.Pattern. It's been that way for years and
+                 * that class has a zillion things on it.
+                 */
+                semanticScope.putDecoration(userRegexNode, new ValueType(Pattern.class));
+                semanticScope.putDecoration(userRegexNode, new StandardConstant(compileJavaPattern(userRegexNode)));
+                return;
+            case GROK:
+                FlavoredPattern grok = new FlavoredPattern.ForGrok(compileGrok(userRegexNode, semanticScope));
+                semanticScope.putDecoration(userRegexNode, new ValueType(FlavoredPattern.class));
+                semanticScope.putDecoration(userRegexNode, new StandardConstant(grok));
+                return;
+            case DISECT:
+                FlavoredPattern dissect = new FlavoredPattern.ForDissect(compileDissect(userRegexNode));
+                semanticScope.putDecoration(userRegexNode, new ValueType(FlavoredPattern.class));
+                semanticScope.putDecoration(userRegexNode, new StandardConstant(dissect));
+                return;
+            default:
+                throw userRegexNode.createError(
+                    new IllegalArgumentException("Unsupported regex flavor [" + userRegexNode.getFlavor() + "]")
+                );
         }
+    }
 
-        Location location = userRegexNode.getLocation();
+    private Pattern compileJavaPattern(ERegex regex) {
+        String pattern = regex.getPattern();
+        String flags = regex.getFlags();
+        try {
+            return Pattern.compile(pattern, convertFlagsToJavaFlags(flags));
+        } catch (PatternSyntaxException pse) {
+            Location errorLocation = new Location(
+                regex.getLocation().getSourceName(),
+                regex.getLocation().getOffset() + regex.getPatternStart() + pse.getIndex()
+            );
+            throw errorLocation.createError(
+                new IllegalArgumentException(
+                    "Could not compile java regex constant [" + pattern + "] with flags [" + flags + "]: " + pse.getDescription(),
```

Comment:
When I was debugging *stuff* I noticed that I wasn't getting good error message from bad java regexes. It turns out that `pse.getDescription()` has the actual cause of the syntax error in it and it isn't included in `toString`! I've added this because I wanted it when I was debugging stuff but it isn't really related to the rest of the change.